cwms-data-api/src/main/java/cwms/cda/data/dao/location/kind/OutletDao.java

```diff
@@ -0,0 +1,229 @@
+            CwmsId projectId = LocationUtil.getLocationIdentifier(outletStruct.getPROJECT_LOCATION_REF());
+
+            Map<String, List<CompoundOutletRecord>> records = new HashMap<>();
+            try {
+                List<CompoundOutletRecord> compoundOutletRecords = retrieveCompoundOutlet(config, locationId,
+                                                                                          projectId.getName(),
+                                                                                          officeId);
+                records.put(locationId, compoundOutletRecords);
+            } catch (RuntimeException e) {
+                LOGGER.atFinest().withCause(e).log("No compound outlet records for outlet " + locationId);
+            }
+
+
+            LocationGroupDao locGroupDao = new LocationGroupDao(dsl);
+            List<LocationGroup> groups = locGroupDao.getLocationGroups(officeId, RATING_LOC_GROUP_CATEGORY);
+
+            return mapToOutlet(outletStruct, records, groups);
+        });
+    }
+
+    private String getRatingGroupId(String locationId, List<LocationGroup> groups) {
+        return groups.stream()
+                     .filter(group -> group.getAssignedLocations()
+                                           .stream()
+                                           .anyMatch(loc -> loc.getLocationId().equalsIgnoreCase(locationId)))
+                     .findFirst()
+                     .map(LocationGroup::getId)
+                     .orElse(null);
+    }
+
+    public void storeOutlet(ProjectStructure outlet, String ratingGroup, boolean failIfExists) {
+        connection(dsl, conn -> {
+            PROJECT_STRUCTURE_OBJ_T structure = mapToProjectStructure(outlet);
+            CWMS_OUTLET_PACKAGE.call_STORE_OUTLET(DSL.using(conn).configuration(), structure, ratingGroup,
+                                                  OracleTypeMap.formatBool(failIfExists));
+        });
+    }
+
+    public void deleteOutlet(String locationId, String officeId, DeleteRule deleteRule) {
+        connection(dsl, conn -> CWMS_OUTLET_PACKAGE.call_DELETE_OUTLET(DSL.using(conn).configuration(), locationId,
+                                                                       deleteRule.getRule(), officeId));
+    }
+
+    private Map<String, List<CompoundOutletRecord>> retrieveCompoundOutletsForProject(Configuration config,
+                                                                                      String projectId,
+                                                                                      String officeId) {
+        Map<String, List<CompoundOutletRecord>> recordMap = new HashMap<>();
+        //projectId and officeId are used as a mask in RETRIEVE_COMPOUND_OUTLETS, however this usage expects a specific
+        //id, since retrieveOutlets does not use a mask.
+        STR_TAB_TAB_T tabs = CWMS_OUTLET_PACKAGE.call_RETRIEVE_COMPOUND_OUTLETS(config, projectId, officeId);
+
+        if (!tabs.isEmpty()) {
```

Comment:
Check size is at least 2 for protection

Reply:
`tabs` is a STR_TAB_TAB_T that will only contain one element.  The loop below is on `tab` which is a STR_TAB_T, which can contain 2 or more elements, and the for loop does that check for us.

Reply:
ah gotcha! thanks